File: python/problem_20.py
class Solution:
    def isValid(self, s: str) -> bool:
        paras = []
        for char in s:
            if char in ["{", "(", "["]:
                paras.append(char)
            else:
                if len(paras) == 0:
                    return False
                if paras[-1] != "{" and char == "}":
                    return False
                if paras[-1] != "[" and char == "]":
                    return False
                if paras[-1] != "(" and char == ")":
                    return False
                paras.pop()

        return len(paras) == 0
                

if __name__ == "__main__":
    s = Solution()
    assert s.isValid("()") == True, "Test case 1 failed"
    assert s.isValid("()[]{}") == True, "Test case 2 failed"
    assert s.isValid("(]") == False, "Test case 3 failed"
    assert s.isValid("((") == False, "Test case 4 failed"

    print("All test cases passed")

File: go/problem_20.go
package main

func isValid(s string) bool {
	paras := []string{}
	for _, char := range s {
		strChar := string(char)
		if strChar == "{" || strChar == "(" || strChar == "[" {
			paras = append(paras, strChar)
		} else {

			if len(paras) == 0 {
				return false
			}

			if strChar == "}" && paras[len(paras)-1] != "{" {
				return false
			}

			if strChar == "]" && paras[len(paras)-1] != "[" {
				return false
			}

			if strChar == ")" && paras[len(paras)-1] != "(" {
				return false
			}

			paras = paras[:len(paras)-1]
		}
	}

	return len(paras) == 0
}
